core/src/main/java/io/parsingdata/metal/data/ImmutableList.java

```diff
@@ -59,9 +63,13 @@ private ImmutableList(final T head, final ImmutableList<T> tail) {
 
     public ImmutableList<T> add(final ImmutableList<T> list) {
         checkNotNull(list, "list");
-        if (list.isEmpty()) { return this; }
```

Comment:
Yes, I saw that it was added in the recursive step. The one reason for my comment is that if we leave the check here, there is no need for a trampoline at all. A _very_ minor optimization.